File: C/UDP-examples/udp-2-server.cc
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <arpa/inet.h>
#include <sys/errno.h>
#include <string.h>

//
// Sample UDP server.
//

int
main(int argc, char **argv)
{
  struct sockaddr_in addr;
  int sockfd;
  int ret;

  fprintf(stdout, "Specified network is *%s*\n", argv[1]);
  fprintf(stdout, "Specified port is *%s*\n", argv[2]);

  sockfd = socket(AF_INET, SOCK_DGRAM, 0);
  if ( sockfd < 0 ) {
    perror("socket");
    exit(1);
  }
  //
  // Initialize socket
  //
  memset(&addr, 0, sizeof(addr));
  addr.sin_family = AF_INET;
  inet_pton(AF_INET, argv[1], &addr.sin_addr);
  addr.sin_port = htons(atoi(argv[2]));

  ret = bind(sockfd, (struct sockaddr *) &addr, sizeof(addr));
  if ( ret < 0 ) {
    perror("bind");
    exit(1);
  }

  for(;;) {
    const int size = 1024;
    char buffer[size];
    struct sockaddr_in from_addr;
    socklen_t from_size = sizeof(from_addr);
    char from_name[1024];

    ret = recvfrom(sockfd, buffer, size, 0,
		   (struct sockaddr*) &from_addr, &from_size);
    if ( ret < 0 ) {
      perror("recvfrom");
      exit(1);
    }

    //
    // Null terminate in case sender didn't or
    // buffer overflow
    //
    buffer[size-1] = 0;
    
    inet_ntop(AF_INET,&from_addr.sin_addr,
	      from_name, sizeof(from_name));

    fprintf(stdout,"Received messsage from %s (port %d): *%s*\n", 
	    from_name, ntohs(from_addr.sin_port), buffer);
  }

  return(0);
}

File: Python/UDP-examples/udp-2-server.py
#!/usr/bin/env python3
##
## Corresponds to udp-2-server.c
##
import socket
import sys,os


##
## Sample UDP server - waits for message, prints it.
## Pair with udp-2-client.c
##

if len(sys.argv) < 3:
    print("usage:", sys.argv[0], "<network> <port>")
    sys.exit(1)

network = sys.argv[1]
port = sys.argv[2]

##
## Create a socket, bind to any interface, specific port
##
sockfd = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sendAddress = (network, int(port))
sockfd.bind(sendAddress)

while True:
    line, addr = sockfd.recvfrom(8192)
    print(f"Server received from {addr[0]} port {addr[1]}: {line.decode('utf-8').rstrip()}")
print("Done")
